pkg/controller/master/storagenetwork/storage_network.go

```diff
@@ -0,0 +1,845 @@
+		}
+
+		// retry again
+		logrus.Warnf("check net-attach-def existing error %v", err)
+		return err
+	}
+
+	if err := h.networkAttachmentDefinitions.Delete(namespace, name, &metav1.DeleteOptions{}); err != nil {
+		logrus.Warnf("remove nad error %v", err)
+		return err
+	}
+
+	setting.Annotations[OldNadStorageNetworkAnnotation] = ""
+	return nil
+}
+
+// return true as finished
+func (h *Handler) checkLonghornSetting(setting *harvesterv1.Setting) bool {
+	value, err := h.getLonghornStorageNetwork()
+	if err != nil {
+		logrus.Warnf("get Longhorn settings error: %v", err)
+		return false
+	}
+
+	currentNad := setting.Annotations[NadStorageNetworkAnnotation]
+	if currentNad == value {
+		// check if we need to restart monitoring pods
+		if !h.checkPodStatusAndStart() {
+			if _, err := h.setConfiguredCondition(setting, false, ReasonInProgress, MsgRestartPod); err != nil {
+				logrus.Warnf("update status error %v", err)
+			}
+			h.settingsController.EnqueueAfter(setting.Name, 5*time.Second)
+			return true
+		}
+
+		settingCopy := setting.DeepCopy()
+		if err := h.removeOldNad(settingCopy); err != nil {
+			logrus.Warnf("remove old nad error %v", err)
+			h.settingsController.EnqueueAfter(setting.Name, 5*time.Second)
+			return true
+		}
+
+		// finish
+		if _, err := h.setConfiguredCondition(settingCopy, true, ReasonCompleted, ""); err != nil {
+			logrus.Warnf("update status error %v", err)
+		}
+		return true
+	}
+	return false
+}
+
+// true: all detach
+func (h *Handler) checkLonghornVolumeDetach() (bool, error) {
+	volumes, err := h.longhornVolumeCache.List(util.LonghornSystemNamespaceName, labels.Everything())
+	if err != nil {
+		logrus.Warnf("volume error %v", err)
+		return false, err
+	}
+
+	for _, volume := range volumes {
+		logrus.Infof("volume state: %v", volume.Status.State)
+		if volume.Status.State != "detached" {
+			return false, nil
+		}
+	}
+	return true, nil
+}
+
+func (h *Handler) checkPrometheusStatusAndStart() bool {
+	// check prometheus cattle-monitoring-system/rancher-monitoring-prometheus replica
+	prometheus, err := h.prometheusCache.Get(CattleMonitoringSystemNamespace, RancherMonitoringPrometheus)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			logrus.Infof("prometheus not found. skip")
+			return true
+		}
+		logrus.Warnf("prometheus get error %v", err)
+		return false
+	}
+
+	logrus.Infof("prometheus: %v", *prometheus.Spec.Replicas)
+	// check started or not
+	if *prometheus.Spec.Replicas == 0 {
+		logrus.Infof("start prometheus")
+		prometheusCopy := prometheus.DeepCopy()
+		*prometheusCopy.Spec.Replicas = 1
+		if replicas, err := strconv.Atoi(prometheus.Annotations[ReplicaStorageNetworkAnnotation]); err == nil {
+			*prometheusCopy.Spec.Replicas = int32(replicas)
+		}
+		delete(prometheusCopy.Annotations, ReplicaStorageNetworkAnnotation)
+
+		if _, err := h.prometheus.Update(prometheusCopy); err != nil {
+			logrus.Warnf("prometheus update error %v", err)
+			return false
+		}
+		return false
```

Comment:
`return false` ? seems should be `true`

Reply:
return false and let controller re-enquque again, and check it again later

Reply:
When everything works well, return `false` is confusing

1. The return result of `h.prometheus.Update` is defined, when OK, it means the updating is successful (the TCP based operation with reply); it is updated to `apiserver`.

2. After 5s to check again, seems not make sense at this point, if you assue that after 5s the POD will be finally stopped, it is not the fact, the POD may take 30s or more to stop (the grace period). You need to list the related POD until len ==0, that means the POD is really gone.  

3. When the return bool means requeue, it's better to define the function as 

`func (h *Handler) checkPrometheusStatusAndStart() (reQueue bool, err error)`, and return `true` when requeue is needed

the named return param will help others to understand

4. In general, it's always essential to return error, for caller to do futher process

Reply:
> After 5s to check again, seems not make sense at this point, if you assue that after 5s the POD will be finally stopped, it is not the fact, the POD may take 30s or more to stop (the grace period). You need to list the related POD until len ==0, that means the POD is really gone.

We didn't expect this.
We will requeue and check repicas and it will skip to next check point which is checking longhorn's volume status

> In general, it's always essential to return error, for caller to do futher process

we don't need error because we log it in function and ignore it.

Reply:
>  When the return bool means requeue, it's better to define the function as

It didn't mean enqueue.
It means check is failed, we need to check it again.

Reply:
> > After 5s to check again, seems not make sense at this point, if you assue that after 5s the POD will be finally stopped, it is not the fact, the POD may take 30s or more to stop (the grace period). You need to list the related POD until len ==0, that means the POD is really gone.
> 
> We didn't expect this. We will requeue and check repicas and it will skip to next check point which is checking longhorn's volume status

The code logic is as following, it assumes that after 5s, `all pods are stopped`, it's not the fact.

`checkPodStatusAndStop` should `NOT` requeue self when update is successful, it makes no sense, because finally, you are relying on the `checkLonghornVolumeDetach`.  In fact, it is `checkLonghornVolumeDetach`  returns false and requeue again and again, unitil all volumes are detached.  This is not a k8s `write` action, but `read`.


```
	// if replica eq 0, skip
	// save replica to annotation
	// set replica to 0
	if !h.checkPodStatusAndStop() {
..
		h.settingsController.EnqueueAfter(settingCopy.Name, 5*time.Second)
..
	}

	logrus.Infof("all pods are stopped")

	// check volume detach before put LH settings
	if ok, err := h.checkLonghornVolumeDetach(); !ok {
...
		h.settingsController.EnqueueAfter(settingCopy.Name, 5*time.Second)
		return nil, nil
	}
```

> 
> > In general, it's always essential to return error, for caller to do futher process
> 
> we don't need error because we log it in function and ignore it.

It's still better to return the error, let caller decides to use or skip.

Reply:
> The code logic is as following, it assumes that after 5s, all pods are stopped, it's not the fact.

Ok, I should move the logrus to another place, sorry to confuse you about that.

Reply:
> checkPodStatusAndStop should NOT requeue self when update is successful, it makes no sense, 

because that is a check, if we check fail, we should re-check it again.

Reply:
> It's still better to return the error, let caller decides to use or skip.

I will consider handling error inside the function because the caller only needs to know success or failure.

Reply:
I will change all return to return true